pkg/client/introduce/client.go

```diff
@@ -0,0 +1,172 @@
+	svc, err := ctx.Service(introduce.Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	introduceSvc, ok := svc.(service.DIDComm)
+	if !ok {
+		return nil, errors.New("cast service to Introduce Service failed")
+	}
+
+	store, err := ctx.StorageProvider().OpenStore(introduce.Introduce)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Client{
+		Event:   introduceSvc,
+		service: introduceSvc,
+		store:   store,
+	}, nil
+}
+
+// SendProposal sends proposal to introducees
+func (c *Client) SendProposal(dest1, dest2 *service.Destination) error {
+	return c.sendProposal(Options{
+		Destinations: []*service.Destination{dest1, dest2},
+	})
+}
+
+// SendProposalWithInvitation sends proposal to introducee
+// executes when the client has a public Invitation
+func (c *Client) SendProposalWithInvitation(inv *didexchange.Invitation, dest *service.Destination) error {
+	return c.sendProposal(Options{
+		Invitation:   inv,
+		Destinations: []*service.Destination{dest},
+	})
+}
+
+// SendRequest sends request
+// sending a request means that introducee is willing to share its invitation
+// the function receives an invitation which can be sent to the introducer
+// an invitation will not be sent if introducer responded with a public invitation
+func (c *Client) SendRequest(invitation *didexchange.Invitation, destination *service.Destination) error {
+	return c.handleOutbound(&introduce.Request{
+		Type: introduce.RequestMsgType,
+		ID:   uuid.New().String(),
+	})
+}
+
+// HandleRequest executes after receiving a request (the client does not have a public Invitation)
+func (c *Client) HandleRequest(rID string, dest1, dest2 *service.Destination) error {
+	return c.sendProposal(Options{
+		Destinations: []*service.Destination{dest1, dest2},
+		ThreadID:     rID,
+	})
+}
+
+// HandleRequestWithInvitation executes after receiving a request when the client has a public Invitation
+func (c *Client) HandleRequestWithInvitation(rID string, inv *didexchange.Invitation, dest *service.Destination) error {
+	return c.sendProposal(Options{
+		Invitation:   inv,
+		Destinations: []*service.Destination{dest},
+		ThreadID:     rID,
+	})
+}
+
+func (c *Client) handleOutbound(msg interface{}) error {
+	payload, err := json.Marshal(msg)
+	if err != nil {
+		return err
+	}
+
+	didMsg, err := service.NewDIDCommMsg(payload)
+	if err != nil {
+		return err
+	}
+
+	return c.service.HandleOutbound(didMsg, &service.Destination{})
+}
+
+// Options keeps the information needed for sending a proposal
+type Options struct {
+	// Invitation must be set when we have a public Invitation
+	Invitation *didexchange.Invitation
+	// Destinations contain one or two elements
+	// one element - for a public Invitation, otherwise two elements
+	Destinations []*service.Destination
+	// ThreadID must be an ID from incoming Request
+	ThreadID string
+}
+
+// sendProposal sends proposal
+// the function may receive an invitation or <nil> (Options.Invitation)
```

Comment:
I had to read this line a few times to grok that you were talking about the Invitation field :).